### PR TITLE
CASSANDRA-17510 streaming three min timeout

### PR DESCRIPTION
**Pull request by jonmeredith**

### CODE REVIEW

File: test/unit/org/apache/cassandra/net/AsyncStreamingInputPlusTest.java [hunk 1]
```diff
@@ -34,8 +34,6 @@
 import io.netty.buffer.Unpooled;
 import io.netty.channel.embedded.EmbeddedChannel;
 import org.apache.cassandra.io.util.BufferedDataOutputStreamPlus;
-import org.apache.cassandra.net.AsyncStreamingInputPlus;
-import org.apache.cassandra.net.AsyncStreamingInputPlus.InputTimeoutException;
 
 import static org.apache.cassandra.utils.Clock.Global.nanoTime;
```
Comment (maedhroz) [inline test/unit/org/apache/cassandra/net/AsyncStreamingInputPlusTest.java]: unused

Comment (maedhroz) [reply test/unit/org/apache/cassandra/net/AsyncStreamingInputPlusTest.java]: also `import java.util.concurrent.TimeUnit;` is unused

File: src/java/org/apache/cassandra/net/AsyncStreamingInputPlus.java [hunk 2]
```diff
@@ -217,23 +209,31 @@ public void close()
...
+                throw new UncheckedInterruptedException(ie);
             }
-        }
+        } while (channel.isRegistered());
 
-        isClosed = true;
+        if (!channel.isRegistered())
+        {
+            throw new IllegalStateException("AsyncStreamingInputPlus did not complete close before channel was unregistered");
+        }
```
Comment (maedhroz) [inline src/java/org/apache/cassandra/net/AsyncStreamingInputPlus.java:212]: nit/style: I don't remember what our style guide says, but I'd be fine with getting rid of the braces for one-liners

File: test/unit/org/apache/cassandra/net/AsyncStreamingInputPlusTest.java [hunk 3]
```diff
@@ -161,6 +129,61 @@ public void available_ClosedButWithBytes()
         Assert.assertEquals(size, inputPlus.unsafeAvailable());
     }
 
+    @Test
+    public void rebufferAndCloseToleratesInterruption() throws InterruptedException
+    {
+        ByteBuf beforeInterrupt = channel.alloc().heapBuffer(1024);
+        beforeInterrupt.writeCharSequence("BEFORE", Charset.forName("us-ascii"));
```
Comment (maedhroz) [inline test/unit/org/apache/cassandra/net/AsyncStreamingInputPlusTest.java]:
```suggestion
        beforeInterrupt.writeCharSequence("BEFORE", StandardCharsets.US_ASCII);
```

File: test/unit/org/apache/cassandra/net/AsyncStreamingInputPlusTest.java [hunk 4]
```diff
@@ -161,6 +129,61 @@ public void available_ClosedButWithBytes()
...
+        ByteBuf afterInterrupt = channel.alloc().heapBuffer(1024);
+        afterInterrupt.writeCharSequence("AFTER", Charset.forName("us-ascii"));
+        final int afterInterruptBytes = afterInterrupt.readableBytes();
+
+        inputPlus = new AsyncStreamingInputPlus(channel);
+        Thread consumer = new Thread(() -> {
+            try
+            {
+                byte[] buffer = new byte[beforeInterruptBytes + afterInterruptBytes];
+                inputPlus.read(buffer, 0, beforeInterruptBytes + afterInterruptBytes);
```
Comment (maedhroz) [inline test/unit/org/apache/cassandra/net/AsyncStreamingInputPlusTest.java]: Would it be worth it to throw a quick check for the number of bytes read?

```
int bytesRead = inputPlus.read(buffer, 0, beforeInterruptBytes + afterInterruptBytes);
assertEquals(buffer.length, bytesRead);
```

**Review by maedhroz — APPROVED** [review]

File: test/unit/org/apache/cassandra/net/AsyncStreamingInputPlusTest.java (hunk 4 above)
Comment (jonmeredith) [reply test/unit/org/apache/cassandra/net/AsyncStreamingInputPlusTest.java]: Can't hurt.

File: test/unit/org/apache/cassandra/net/AsyncStreamingInputPlusTest.java (hunk 3 above)
Comment (jonmeredith) [reply test/unit/org/apache/cassandra/net/AsyncStreamingInputPlusTest.java]: Much better.

**Comment by jonmeredith** [comment]
Merged

**Pull request closed without merging**